member/src/main/scala/com/improving/app/member/domain/Member.scala

```diff
@@ -56,220 +68,409 @@ object Member extends StrictLogging {
+        tenant = editInfo.tenant.orElse(info.tenant)
       )
+
+      val newMeta = metaInfo.copy(
+        lastModifiedBy = editMemberInfoCommand.editingMember,
+        lastModifiedOn = Some(Timestamp(Instant.now(clock)))
+      )
+      val event = MemberInfoEdited(editMemberInfoCommand.memberId, Some(newInfo), Some(newMeta))
+      Effect.persist(event).thenReply(replyTo) { _ => StatusReply.Success(MemberEventResponse(event)) }
     }
   }
 
-  private def createMemberMetaInfo(createdBy: MemberId, createdOn: Timestamp): MemberMetaInfo =
-    MemberMetaInfo(
-      Some(createdOn),
-      Some(createdBy),
-      Some(createdOn),
-      Some(createdBy),
-      MemberStatus.MEMBER_STATUS_INITIAL
-    )
-
-  //Will fail if invalid state
-  private def updateMemberMetaInfo(
-      state: MemberState,
-      updatedBy: MemberId,
-      updatedOn: Timestamp,
-      updatedStatus: MemberStatus
-  ): MemberState =
-    state.withMemberMetaInfo(
-      state.memberMetaInfo.get
-        .withLastModifiedBy(updatedBy)
-        .withLastModifiedOn(updatedOn)
-        .withMemberState(updatedStatus)
-    )
-
-  //EventHandler
-  private val eventHandler: (MemberState, MemberEvent) => MemberState = { (state, event) =>
-    event match {
-
-      case MemberRegistered(_, Some(memberInfo), Some(actingMember), Some(createdOn), _) =>
-        state.withMemberInfo(memberInfo).withMemberMetaInfo(createMemberMetaInfo(actingMember, createdOn))
+  private def editMemberInfo(
+    state: MemberState,
+    editMemberInfoCommand: EditMemberInfo,
+    replyTo: ActorRef[StatusReply[MemberResponse]]
+  ): ReplyEffect[MemberEvent, MemberState] = {
+    MemberValidation.validateEditMemberInfo(editMemberInfoCommand) match {
+      case Valid(_) =>
+        state match {
+          case DraftMemberState(requiredInfo, optionalInfo, meta) =>
+            val info = createMemberInfoFromDraftState(requiredInfo, optionalInfo)
+            editMemberInfoLogic(info, meta, editMemberInfoCommand, replyTo)
+          case RegisteredMemberState(info, meta) => editMemberInfoLogic(info, meta, editMemberInfoCommand, replyTo)
+          case _: TerminatedMemberState => useTerminatedStateErrorStatusReply(replyTo)
+        }
+      case Invalid(errors) =>
+        useCommandValidationErrorStatusReply(replyTo, errors, "EditMemberInfo")
+    }
+  }
 
-      case MemberActivated(_, Some(actingMember), Some(updatedOn), _) =>
-        updateMemberMetaInfo(state, actingMember, updatedOn, MEMBER_STATUS_ACTIVE)
+  private def getMemberInfoLogic (
+    info: MemberInfo,
+    metaInfo: MemberMetaInfo,
+    getMemberInfoCommand: GetMemberInfo,
+    replyTo: ActorRef[StatusReply[MemberResponse]]
+  ): ReplyEffect[MemberEvent, MemberState] = {
+    val (infoOpt, metaOpt): (Option[MemberInfo], Option[MemberMetaInfo]) =
+      metaInfo.createdBy.fold[(Option[MemberInfo], Option[MemberMetaInfo])]((None, None)) {
```

Comment:
This doesn't seem like it belongs here... why are you folding on `createdBy`?

Reply:
The fold is needed here because we want to return None for both memberInfo and member meta if the member has not been registered (this is part of the specifications/riddl). It's just another way of saying 

if (metaInfo.createdBy.isEmpty) //return None and None else //return the memberInfo and metaInfo

Reply:
I find that fold tends to obscure the intent of code, and so if I find it is necessary, I like to wrap it in a descriptively named function to help clarify the intent.

Reply:
The other thing of note here is that the presence of the createdBy is standing in for something else here. I would prefer to see whatever that is be made explicit (and while we're at it, I bet we can restructure this and set up the state types to get a compile-time guarantee that we only get to this code if the state is correct)

Reply:
Dag got there before me. My issue is exactly that `createdBy` is standing in for the entire metaInfo object - I would prefer not to use a random subfield but rather the `metaInfo` itself.